Export logs

wbobeirne opened the pull request:
Closes #1109.

### What This Does

Adds a share button to the top right of the logs on mobile that loads the log full log file and pops up the device's native share dialog to be able to send it somewhere else (Most likely to save it as a file or send it in an email.)

For now, the button just doesn't show on the desktop electron app, but maybe a good follow up would be to have this button open the folder the log file is in so that they could easily upload it.

### Steps to Test

1. Open the app on mobile and go to your logs, confirm you see a share button in the top right
2. Press the share button, confirm you are able to export the logs to your preferred app
3. Open the app on desktop and go to your logs, confirm you do not see a share button in the top right

## Screenshots

![log-export](https://user-images.githubusercontent.com/649992/61483334-98ff3380-a96a-11e9-86d7-f774c9c19f45.png)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/lightninglabs/lightning-app?pullRequest=1267) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/lightninglabs/lightning-app?pullRequest=1267) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/lightninglabs/lightning-app?pullRequest=1267) it.</sub>

wbobeirne commented on src/action/log.js:
```diff
@@ -49,6 +50,29 @@ export function error(...args) {
+  }
+  return _FS.readFile(getLogPath(_store.network), 'utf8');
+}
+
```
Added some tests with and without the required constructor args. Let me know if the way I stubbed RNFS and Share was what you normally do.

wbobeirne commented on src/view/cli.js:
```diff
@@ -19,16 +20,37 @@ const styles = StyleSheet.create({
+    } catch (err) {
+      error(err);
+    }
+  }
```
Good call, moved it out to a `shareLogs` function exported from `actions/log.js`

tanx replied:
As mentioned above, we should move this try/catch block into `shareLogs` itself and call `onPress={() => log.shareLogs()}`. Then we can keep the react component purely functional as before and prevent breakout into the `class` syntax.

tanx commented on src/action/log.js:
```diff
@@ -49,6 +51,49 @@ export function error(...args) {
+  if (!_FS) {
+    return Promise.reject(
+      new Error('Cannot get logs with no FS in action/log.js')
+    );
```
If we declare `async function getLogs` we be can throw an Error here directly. No need to wrap in `Promise.reject()` in that case.

tanx commented on src/action/log.js:
```diff
@@ -49,6 +51,49 @@ export function error(...args) {
+
+/**
+ * Retrieves the entire LND log file as a string.
+ * @return {Promise}
```
`@return {Promise<string>}`

tanx commented on src/action/log.js:
```diff
@@ -49,6 +51,49 @@ export function error(...args) {
+  if (!_Share) {
+    return Promise.reject(
+      new Error('Cannot share logs with no Share in action/log.js')
+    );
```
Same here ... no need to wrap in `Promise.reject` due to `async function shareLogs`. Just do a `throw new Error(...)`.

tanx commented on src/action/log.js:
```diff
@@ -49,6 +51,49 @@ export function error(...args) {
+  return _Share.share({
+    title: 'Lightning App logs',
+    message: logs,
+  });
```
We should wrap the complete contents of `shareLogs` in a try/catch and log the error here with `this.error(err)`. Then we don't need to wrap the function call in a try/catch in the view. That way we could keep the view react component purely functional.

wbobeirne replied:
Are there any instances in which we may want to have custom logic when it throws? (E.g. popping up an error message to the user?) I typically avoid eating errors in case it needs to be handled.

tanx commented on src/action/log.js:
```diff
@@ -59,10 +104,12 @@ function pushLogs(message) {
 
 class LogAction {
-  constructor(store, ipc, printErrObj = true) {
+  constructor(store, ipc, printErrObj = true, FS, Share) {
```
👍

tanx commented on src/view/cli.js:
```diff
@@ -19,16 +20,33 @@ const styles = StyleSheet.create({
   },
 });
 
-const CLIView = ({ store, nav }) => (
```
We should inject the `log` instance of the LogActions as a prop into this view via `({ store, nav, log })`. And inject `log` in https://github.com/lightninglabs/lightning-app/blob/1ee33ed1651bd84a87174a5ebc3c16b27c0e030f/src/view/main-mobile.js#L140

wbobeirne replied:
Right now all of the log action's functions are standalone and individually exported & imported. Should they all be moved into the Log class and called like that?

tanx commented on test/unit/action/log.spec.js:
```diff
@@ -10,6 +10,14 @@ describe('Action Logs Unit Tests', () => {
+  const Share = {
+    share: () => Promise.resolve(true),
+  };
+
```
Stubs should be declared via `let` here and initialized using `sinon.stub().resolves(true)`  in the `beforeEach` block below. That way all test case can work with a new stub instance and check how often it was called.

wbobeirne commented:
* Removed unnecessary `Promise.reject`'s
* Made `shareLogs` a self-handling component so that it doesn't require catching errors
* Added sinon stubs to tests and check that the relevant functions get called

The only outstanding issue is the passing of `log` as a prop, let me know how you feel about that one and if it needs changes.

tanx commented:
> * Removed unnecessary `Promise.reject`'s
> * Made `shareLogs` a self-handling component so that it doesn't require catching errors
> * Added sinon stubs to tests and check that the relevant functions get called

@wbobeirne awesome! Thanks for the updates.

> The only outstanding issue is the passing of `log` as a prop, let me know how you feel about that one and if it needs changes.

Indeed. It doesn't seem to be so simple since we don't actually use this action that way for the other functions. I can look into this. No need to dig further into it on your end.

Thanks so much for the PR 👍

tanx commented:
@wbobeirne I went ahead and moved the file I/O operations into its own action module. That allowed us to inject `file` via a prop into the cli view like the other actions. Other than that great work!

Closing this PR in favor of https://github.com/lightninglabs/lightning-app/pull/1270
